jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/ComplianceViolation.java

```diff
@@ -75,13 +81,169 @@ interface Mode
         Set<? extends ComplianceViolation> getAllowed();
     }
 
+    public static record Event(ComplianceViolation.Mode mode, ComplianceViolation violation, String details)
+    {
+        @Override
+        public String toString()
+        {
+            return String.format("%s (see %s) in mode %s for %s",
+                violation.getDescription(), violation.getURL(), mode, details);
+        }
+    };
+
     /**
      * A listener that can be notified of violations.
      */
     interface Listener
     {
+        /**
+         * A new Request has begun.
+         *
+         * @param request the request attributes, or null if the Request does not exist yet (eg: during parsing of HTTP/1.1 headers, before request is created)
+         */
+        default void onRequestBegin(Attributes request)
```

Comment:
It would make more sense to me that this method returned a Listener to be used for all future compliance events with the request.   Then a capturing listener would just return a new instance here without the need for a specific factory.

ah but we might have compliance events before the onRequestBegin.   So we just need a new factory method to ask the listener for an instance as soon as we start a request cycle.  it can return this if it is not capturing or a new instance if it is.